.travis.yml

```diff
@@ -4,42 +4,77 @@ cache:
   directories:
     - "$HOME/.npm"
 
-node_js:
-  - '6'
-  - '7'
-  - '8'
-
-env:
-  - MODULE="quantum-api"
-  - MODULE="quantum-code-highlight"
-  - MODULE="quantum-core"
-  - MODULE="quantum-diagram"
-  - MODULE="quantum-docs"
-  - MODULE="quantum-dom"
-  - MODULE="quantum-html"
-  - MODULE="quantum-template"
-  - MODULE="quantum-markdown"
-  - MODULE="quantum-version"
-  - MODULE="integration"
-  # - MODULE="docs"
-
-before_install:
-  - if [ "$MODULE" != "integration" ]; then cd "$MODULE" || return; fi
-
-install:
-  - if [ "$MODULE" != "integration" ]; then npm install; else npm install && npm run link-all; fi
-
-script:
-  - if [ "$MODULE" != "integration" ]; then npm test && npm run upload-coverage; else npm test; fi
-
-deploy:
-  email: quantumjs.ci.noreply@ocado.com
-  provider: npm
-  api_key:
-    secure: oLEt0deDskq9/dbmOH3KqE7QnubsEVzJz6CCk7q4DUq7O7CoZdXtmG9HOf8TzFlvsFJRrmKtYYA7gZW8/iOlruYSBnJQDi00qr46M4KVxDmBB+uFrZYVQCIyq5t3lHqL7RPjyu5TV8c8TajWQ48RLSET34DDEe1lAQQykBBh/nbM28ZPWmb5DzVpfA+wpzg8YwNPzz2V536A03OVtS9A86HCvkk1EUzz4ldakNsF7HnTmyZK3JjEijIBAkfp+fvUw0ipJMXIOl4cltdwtylSg2sP2qjKJFniWWdD9/4qXNJL4WmPRZ/gmuFxopejcklKClQCl0O313bB5WL8CoSo+EWzmZVx7E6VYmh95cOcVcwKOe2mrVYUmBONj5eUSzX/b9YX3tUIXh3hySSUgyBtZD1yxvPCeM5kvA9mb+lLrio3NVD2SzT0jTEVco40te8vm51DR+OczJz1EiNFqUcM6yCdIny/YDs+KZp3YUP9BfJEscgKZgNkc7reGxqirzzgyomFinFJub4DRQVNpgoVcc7aA1QtAnqGOkYTCQM9Z8mWGnBXOiGpIFD7njBPmRGmLpGE/26T1lJ8HgmCF7ymYVJEcKOMq0eyhFvwMwu1dYOyGLMQnyKZ2TKHzemiUDq1Vtror5La/ytQL0xJLHqe1BSg1cGa2TSHGnptUVZ15/0=
-  on:
-    repo: ocadotechnology/quantumjs
-    branch: master
-    tags: true
-    node: '6'
-    condition: "$TRAVIS_TAG = $MODULE*"
+stages:
+  - test
+
+  - name: publish module
+    if: tag IS present AND repo = ocadotechnology/quantumjs
+
+
+module_test: &module_test
+  stage: test
+  before_install:
+    - cd "$MODULE"
+  install:
+    - npm install
+  script:
+    - npm test && npm run upload-coverage
+  node_js: '8'
+
+integration_test: &integration_test
+  stage: test
+  env:
+    - MODULE="integration"
+  install:
+    - npm install && npm run link-all
+  script:
+    - npm test
+
+jobs:
+  include:
+  - node_js: '6'
+    <<: *integration_test
+
+  - node_js: '8'
+    <<: *integration_test
```

Comment:
Start the integration tests first as they take the longest. This leaves 3 runners to work through the remaining jobs.

Each module test takes ~1 minute (10 modules) and each integration test takes around 5 minutes, providing the job doesn't hang (as it tends to for node 6)

This means that by the time the integration tests complete, the module tests should already have completed - This means that the integration tests dictate the length of time the 'Test' stage will last

Reply:
Seems reasonable. With this setup there is also the option of adding a stages clause that runs the unit tests first, then the integration tests, which would mean the tests fail faster when there are errors. Might mean they take more time overall, so what you've done for now seems good to me

Reply:
I did initially have it like that but I decided against it as it's making the 'happy path' slower to make the 'error path' easier and I figured that didn't make sense as a pattern - Happy to change it though moving forward if waiting for the integration tests becomes too tedious